app/models/user.rb

```diff
@@ -8,4 +8,18 @@ class User < ApplicationRecord
   validates :first_name, presence: true
   validates :last_name, presence: true
   validates :email, presence: true
+  validates_format_of :username, with: /^[a-zA-Z0-9_\.]*$/, multiline: :true
+
+  attr_accessor :login
+
+  def self.find_for_database_authentication(warden_conditions)
+    conditions = warden_conditions.dup
+    if login = conditions.delete(:login)
```

Comment:
Agree with hound - can you get rid of `login`? So `if conditions.delete(:login)`